samcli/lib/telemetry/metric.py

```diff
@@ -0,0 +1,355 @@
+"""
+Provides methods to generate and send metrics
+"""
+from timeit import default_timer
+from functools import wraps, reduce
+
+import uuid
+import platform
+import logging
+import click
+
+from samcli import __version__ as samcli_version
+from samcli.cli.context import Context
+from samcli.cli.global_config import GlobalConfig
+from samcli.lib.warnings.sam_cli_warning import TemplateWarningsChecker
+from samcli.commands.exceptions import UserException
+from .telemetry import Telemetry
+
+LOG = logging.getLogger(__name__)
+
+WARNING_ANNOUNCEMENT = "WARNING: {}"
+
+"""
+Global variables are evil but this is a justified usage.
+This creates a versitile telemetry tracking no matter where in the code. Something like a Logger.
+No side effect will result in this as it is write-only for code outside of telemetry.
+Decorators should be used to minimize logic involving telemetry.
+"""
+_METRICS = dict()
+
+
+def send_installed_metric():
+    LOG.debug("Sending Installed Metric")
+
+    telemetry = Telemetry()
+    metric = Metric("installed")
+    metric.add_data("osPlatform", platform.system())
+    metric.add_data("telemetryEnabled", bool(GlobalConfig().telemetry_enabled))
+    telemetry.emit(metric, force_emit=True)
+
+
+def track_template_warnings(warning_names):
+    """
+    Decorator to track when a warning is emitted. This method accepts name of warning and executes the function,
+    gathers all relevant metrics, reports the metrics and returns.
+
+    On your warning check method use as follows
+
+        @track_warning(['Warning1', 'Warning2'])
+        def check():
+            return True, 'Warning applicable'
+    """
+
+    def decorator(func):
+        """
+        Actual decorator method with warning names
+        """
+
+        def wrapped(*args, **kwargs):
+            telemetry = Telemetry()
```

Comment:
Shouldn't we check here that Telemetry is enabled so that we don't execute all this code if it isn't?

Maybe add a `is_enabled` function to Telemetry and use it

Reply:
The check for telemetry enabled or not is done in Telemetry.emit(...). Building a Metric object is trivial amount of work and having a single point of checking reduces possible error.